File: tests/dart/serialization_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:test/test.dart';
import 'package:path/path.dart' as path;
import '../../clients/cardscan-dart/lib/cardscan_client.dart';

/// Load fixture files from shared test directory
String loadFixture(String filename) {
  final currentDir = Directory.current.path;
  final fixturePath = path.normalize(path.join(currentDir, '..', '..', 'tests', 'fixtures', 'api_responses', filename));
  final file = File(fixturePath);
  
  if (!file.existsSync()) {
    throw Exception('Fixture file not found: ${file.path}');
  }
  
  return file.readAsStringSync();
}

/// Comprehensive Dart serialization tests
/// Tests the Dart client's ability to handle various JSON response formats
void main() {
  group('Dart Client Serialization Tests', () {
    
    group('ScanMetadataCaptureScoreScoresInner Serialization', () {
      test('should handle string numeric values from API', () {
        print('🧪 Testing ScanMetadataCaptureScoreScoresInner with string inputs');
        
        // Simulate the API JSON response with string numeric values
        final apiJsonResponse = {
          'score': '0.994',      // String instead of number
          'laplacian': '0.832'   // String instead of number
        };
        
        print('   Input JSON: $apiJsonResponse');
        print('   - score type: ${apiJsonResponse['score'].runtimeType}');
        print('   - laplacian type: ${apiJsonResponse['laplacian'].runtimeType}');
        
        try {
          // Try to deserialize using the built_value serializer
          final result = standardSerializers.deserializeWith(
            ScanMetadataCaptureScoreScoresInner.serializer, 
            apiJsonResponse
          )!;
          
          print('✅ SUCCESS: Deserialization worked!');
          print('   Result: score=${result.score} (${result.score.runtimeType}), laplacian=${result.laplacian} (${result.laplacian.runtimeType})');
          
          // Verify the values are correctly parsed as numbers
          expect(result.score, isA<num>());
          expect(result.laplacian, isA<num>());
          expect(result.score, equals(0.994));
          expect(result.laplacian, equals(0.832));
          
        } catch (e, stackTrace) {
          print('❌ FAILED: Deserialization failed');
          print('   Error: $e');
          print('   Stack trace: $stackTrace');
          rethrow;
        }
      });

      test('should handle mixed string/numeric JSON', () {
        print('🌍 Testing with mixed string/numeric JSON');
        
        final mixedJson = {
          'score': 0.5,          // Actual number
          'laplacian': '0.75'    // String number
        };
        
        print('   Input JSON: $mixedJson');
        
        try {
          final result = standardSerializers.deserializeWith(
            ScanMetadataCaptureScoreScoresInner.serializer, 
            mixedJson
          )!;
          
          print('✅ SUCCESS: Mixed types handled correctly!');
          print('   Result: score=${result.score} (${result.score.runtimeType}), laplacian=${result.laplacian} (${result.laplacian.runtimeType})');
          
          expect(result.score, equals(0.5));
          expect(result.laplacian, equals(0.75));
          
        } catch (e) {
          print('❌ FAILED: $e');
          rethrow;
        }
      });

      test('should handle purely numeric JSON', () {
        final numericJson = {
          'score': 0.999,
          'laplacian': 0.123
        };
        
        final result = standardSerializers.deserializeWith(
          ScanMetadataCaptureScoreScoresInner.serializer, 
          numericJson
        )!;
        
        expect(result.score, equals(0.999));
        expect(result.laplacian, equals(0.123));
        expect(result.score, isA<num>());
        expect(result.laplacian, isA<num>());
      });
    });

    group('CardApiResponse Serialization', () {
      test('should parse card response from shared fixture', () {
        // Using the same fixtures as other clients for consistency
        final cardJson = {
          'card_id': 'c53bcd29-26ec-4121-b090-cd381cae745f',
          'state': 'pending',
          'created_at': '2025-06-18 02:26:24.578379+00:00',
          'deleted': false
        };
        
        try {
          final result = standardSerializers.deserializeWith(
            CardApiResponse.serializer, 
            cardJson
          )!;
          
          expect(result.cardId, equals('c53bcd29-26ec-4121-b090-cd381cae745f'));
          expect(result.state, equals(CardState.pending));
          expect(result.deleted, equals(false));
          expect(result.createdAt, isNotNull);
          
        } catch (e) {
          print('Card parsing failed: $e');
          rethrow;
        }
      });

      test('should parse completed card with payer match', () {
        final cardWithPayerJson = {
          'card_id': 'c1b93738-ddc0-4beb-9936-1f93fe0e4279',
          'state': 'completed',
          'created_at': '2025-06-18 02:26:24.578379+00:00',
          'deleted': false,
          'payer_match': {
            'cardscan_payer_id': 'pay_8otorlr4',
            'cardscan_payer_name': 'UNITEDHEALTHCARE',
            'score': '0.95'
          }
        };
        
        try {
          final result = standardSerializers.deserializeWith(
            CardApiResponse.serializer, 
            cardWithPayerJson
          )!;
          
          expect(result.cardId, equals('c1b93738-ddc0-4beb-9936-1f93fe0e4279'));
          expect(result.state, equals(CardState.completed));
          expect(result.payerMatch, isNotNull);
          expect(result.payerMatch!.cardscanPayerName, equals('UNITEDHEALTHCARE'));
          expect(result.payerMatch!.score, equals('0.95'));
          
        } catch (e) {
          print('Payer match parsing failed: $e');
          rethrow;
        }
      });
    });

    group('EligibilityApiResponse Serialization', () {
      test('should parse eligibility response', () {
        final eligibilityJson = {
          'eligibility_id': '93376802-779b-42ad-bfa3-d6e99d5a02c9',
          'state': 'processing',
          'card_id': '529e865d-78c2-4f9e-aa9a-addedf642c88',
          'created_at': '2025-06-18 02:26:24.578379+00:00'
        };
        
        try {
          final result = standardSerializers.deserializeWith(
            EligibilityApiResponse.serializer, 
            eligibilityJson
          )!;
          
          expect(result.eligibilityId, equals('93376802-779b-42ad-bfa3-d6e99d5a02c9'));
          expect(result.state.toString().split('.').last, equals('processing'));
          expect(result.cardId, equals('529e865d-78c2-4f9e-aa9a-addedf642c88'));
          
        } catch (e) {
          print('Eligibility parsing failed: $e');
          rethrow;
        }
      });
    });

    group('Error Response Serialization', () {
      test('should parse error card response', () {
        final errorJson = {
          'card_id': 'b7012e64-24c6-4f85-8410-adf36fe03e8a',
          'state': 'error',
          'created_at': '2025-06-18 02:26:24.578379+00:00',
          'deleted': false,
          'error': {
            'type': 'RejectedCard',
            'message': 'Rejecting scan for invalid card',
            'code': 'UKN'
          }
        };
        
        try {
          final result = standardSerializers.deserializeWith(
            CardApiResponse.serializer, 
            errorJson
          )!;
          
          expect(result.cardId, equals('b7012e64-24c6-4f85-8410-adf36fe03e8a'));
          expect(result.state, equals(CardState.error));
          expect(result.error, isNotNull);
          expect(result.error!.type, equals('RejectedCard'));
          expect(result.error!.message, equals('Rejecting scan for invalid card'));
          
        } catch (e) {
          print('Error response parsing failed: $e');
          rethrow;
        }
      });
    });

    group('Field Name Compatibility', () {
      test('should handle snake_case field names from API', () {
        // Test that models can handle snake_case field names (API format)
        final snakeCaseJson = {
          'card_id': 'test-123',
          'state': 'pending',
          'created_at': '2023-01-01T00:00:00Z',
          'deleted': false,
          'payer_match': {
            'cardscan_payer_name': 'TEST_PAYER',
            'score': '0.88'
          }
        };
        
        try {
          final result = standardSerializers.deserializeWith(
            CardApiResponse.serializer, 
            snakeCaseJson
          )!;
          
          expect(result.cardId, equals('test-123'));
          expect(result.createdAt, isNotNull);
          if (result.payerMatch != null) {
            expect(result.payerMatch!.cardscanPayerName, equals('TEST_PAYER'));
          }
          
        } catch (e) {
          print('Snake case field parsing failed: $e');
          rethrow;
        }
      });
    });

    group('Enum Serialization', () {
      test('should correctly parse CardState enum values', () {
        final states = ['pending', 'processing', 'completed', 'error'];
        
        for (final stateString in states) {
          final json = {
            'card_id': 'test-id',
            'state': stateString,
            'created_at': '2023-01-01T00:00:00Z',
            'deleted': false
          };
          
          final result = standardSerializers.deserializeWith(
            CardApiResponse.serializer, 
            json
          )!;
          expect(result.state.toString().split('.').last, equals(stateString));
        }
      });

      test('should correctly parse EligibilityState enum values', () {
        final states = ['processing', 'completed', 'error'];
        
        for (final stateString in states) {
          final json = {
            'eligibility_id': 'test-id',
            'state': stateString,
            'card_id': 'card-123',
            'created_at': '2023-01-01T00:00:00Z'
          };
          
          final result = standardSerializers.deserializeWith(
            EligibilityApiResponse.serializer, 
            json
          )!;
          expect(result.state.toString().split('.').last, equals(stateString));
        }
      });
    });

    group('Null and Optional Field Handling', () {
      test('should handle optional fields gracefully', () {
        final minimalJson = {
          'card_id': 'test-123',
          'state': 'pending',
          'created_at': '2023-01-01T00:00:00Z',
          'deleted': false
          // No optional fields like payer_match, details, etc.
        };
        
        final result = standardSerializers.deserializeWith(
          CardApiResponse.serializer, 
          minimalJson
        )!;
        
        expect(result.cardId, equals('test-123'));
        expect(result.state, equals(CardState.pending));
        expect(result.payerMatch, isNull);
        expect(result.details, isNull);
        expect(result.error, isNull);
      });
    });

    group('Comprehensive Real Fixture Tests', () {
      test('should parse massive card response from payer match fixture', () {
        print('📁 Testing comprehensive card response from real fixture');
        
        final fixtureContent = loadFixture('card_response_with_payer_match.json');
        final fixtureJson = json.decode(fixtureContent) as Map<String, dynamic>;
        
        final result = standardSerializers.deserializeWith(
          CardApiResponse.serializer, 
          fixtureJson
        )!;
        
        // Test basic card info
        expect(result.cardId, equals('c1b93738-ddc0-4beb-9936-1f93fe0e4279'));
        expect(result.state, equals(CardState.completed));
        expect(result.deleted, equals(false));
        
        // Test rich details structure
        expect(result.details, isNotNull);
        
        // Test payer match comprehensive data
        final payerMatch = result.payerMatch!;
        expect(payerMatch.cardscanPayerId, equals('pay_8otorlr4'));
        expect(payerMatch.cardscanPayerName, equals('UNITEDHEALTHCARE'));
        expect(payerMatch.score, equals('0.952'));
        
        // Test metadata
        expect(result.metadata, isNotNull);
        
        // Test images
        expect(result.images, isNotNull);
        expect(result.images!.front, isNotNull);
        
        print('✅ Comprehensive fixture test passed');
        print('   Card ID: ${result.cardId}');
        print('   Payer: ${payerMatch.cardscanPayerName}');
        print('   Score: ${payerMatch.score}');
      });
      
      test('should parse card response with backside from fixture', () {
        print('🖼️ Testing card response with backside from fixture');
        
        final fixtureContent = loadFixture('card_response_with_backside.json');
        final fixtureJson = json.decode(fixtureContent) as Map<String, dynamic>;
        
        final result = standardSerializers.deserializeWith(
          CardApiResponse.serializer, 
          fixtureJson
        )!;
        
        expect(result.cardId, equals('e3f2a892-b360-4aaf-908e-25a12878da1c'));
        expect(result.state, equals(CardState.completed));
        
        // Test both front and back images
        expect(result.images, isNotNull);
        expect(result.images!.front, isNotNull);
        expect(result.images!.back, isNotNull);
        
        final frontUrl = result.images!.front!.url!;
        final backUrl = result.images!.back!.url!;
        expect(frontUrl, contains('https://'));
        expect(backUrl, contains('https://'));
        expect(frontUrl, contains('cardscan-sandbox-uploads'));
        expect(backUrl, contains('cardscan-sandbox-uploads'));
        
        print('✅ Backside fixture test passed');
        print('   Front image: Present');
        print('   Back image: Present');
      });
      
      test('should parse error response from fixture', () {
        print('⚠️  Testing error card response from fixture');
        
        final fixtureContent = loadFixture('card_response_error.json');
        final fixtureJson = json.decode(fixtureContent) as Map<String, dynamic>;
        
        final result = standardSerializers.deserializeWith(
          CardApiResponse.serializer, 
          fixtureJson
        )!;
        
        expect(result.cardId, equals('b7012e64-24c6-4f85-8410-adf36fe03e8a'));
        expect(result.state, equals(CardState.error));
        expect(result.error, isNotNull);
        expect(result.error!.type, equals('RejectedCard'));
        expect(result.error!.message, equals('Rejecting scan for invalid card'));
        
        print('✅ Error response fixture test passed');
      });
      
      test('should parse eligibility response from fixture', () {
        print('🔍 Testing eligibility response from fixture');
        
        final fixtureContent = loadFixture('eligibility_response_processing.json');
        final fixtureJson = json.decode(fixtureContent) as Map<String, dynamic>;
        
        final result = standardSerializers.deserializeWith(
          EligibilityApiResponse.serializer, 
          fixtureJson
        )!;
        
        expect(result.eligibilityId, equals('93376802-779b-42ad-bfa3-d6e99d5a02c9'));
        expect(result.state, equals(EligibilityApiResponseState.processing));
        expect(result.cardId, equals('529e865d-78c2-4f9e-aa9a-addedf642c88'));
        
        print('✅ Eligibility response fixture test passed');
      });
    });
  });
}

File: tests/kotlin/SerializationTest.kt
/**
 * Comprehensive Kotlin serialization tests
 * Tests the Kotlin client's ability to handle various JSON response formats
 * Focuses on string numeric values and field conversion robustness
 */
package org.openapitools.client.tests

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import org.openapitools.client.models.*
import com.squareup.moshi.Moshi
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory

class SerializationTest {
    
    private val moshi = Moshi.Builder()
        .add(KotlinJsonAdapterFactory())
        .build()
    
    private val cardApiResponseAdapter: JsonAdapter<CardApiResponse> = 
        moshi.adapter(CardApiResponse::class.java)
    
    private val eligibilityApiResponseAdapter: JsonAdapter<EligibilityApiResponse> = 
        moshi.adapter(EligibilityApiResponse::class.java)

    @Test
    fun testStringNumericValuesEdgeCase() {
        println("🧪 Testing string numeric values edge case")
        
        // Simulate API response with string numeric values instead of actual numbers
        val apiJsonWithStringNumbers = """
        {
            "card_id": "test-123",
            "state": "completed", 
            "created_at": "2025-06-18T02:26:24.578379+00:00",
            "deleted": false,
            "payer_match": {
                "cardscan_payer_name": "UNITEDHEALTHCARE",
                "bin_number": "610020",
                "issuer_name": "UNITEDHEALTHCARE",
                "score": "0.994"
            }
        }
        """.trimIndent()
        
        println("   Input score: \"0.994\" (string)")
        
        try {
            val cardResponse = cardApiResponseAdapter.fromJson(apiJsonWithStringNumbers)
            assertNotNull(cardResponse, "Card response should not be null")
            
            println("✅ SUCCESS: String number parsed correctly")
            
            // Verify the score is preserved as a string
            assertEquals("test-123", cardResponse!!.cardId, "Card ID should match")
            assertEquals(CardState.completed, cardResponse.state, "State should be completed")
            assertEquals("0.994", cardResponse.payerMatch?.score, "Score should be preserved as string")
            
            println("   Result score: ${cardResponse.payerMatch?.score} (${cardResponse.payerMatch?.score?.javaClass?.simpleName})")
            
        } catch (e: Exception) {
            println("❌ FAILED: String number parsing failed: ${e.message}")
            throw e
        }
    }

    @Test
    fun testMixedStringAndNumericValues() {
        println("🌍 Testing mixed string/numeric JSON")
        
        val mixedJson = """
        {
            "card_id": "test-456",
            "state": "completed",
            "created_at": "2025-06-18T02:26:24.578379+00:00", 
            "deleted": false,
            "payer_match": {
                "cardscan_payer_name": "ANTHEM",
                "bin_number": "987654",
                "issuer_name": "ANTHEM",
                "score": "0.5"
            }
        }
        """.trimIndent()
        
        println("   payer_match score: \"0.5\" (string)")
        
        try {
            val cardResponse = cardApiResponseAdapter.fromJson(mixedJson)
            assertNotNull("Card response should not be null", cardResponse)
            
            println("✅ SUCCESS: String score handled correctly!")
            
            // Verify score is preserved as string
            assertEquals("Score should be string", "0.5", cardResponse!!.payerMatch?.score)
            
            println("   Score preserved as string: ${cardResponse.payerMatch?.score}")
            
        } catch (e: Exception) {
            println("❌ FAILED: Mixed type parsing failed: ${e.message}")
            throw e
        }
    }

    @Test  
    fun testInvalidStringInScoreField() {
        println("⚠️  Testing invalid string in score field")
        
        val malformedJson = """
        {
            "card_id": "test-789",
            "state": "completed",
            "created_at": "2025-06-18T02:26:24.578379+00:00",
            "deleted": false,
            "payer_match": {
                "cardscan_payer_name": "CIGNA",
                "bin_number": "555555", 
                "issuer_name": "CIGNA",
                "score": "invalid_number"
            }
        }
        """.trimIndent()
        
        println("   Invalid score: \"invalid_number\"")
        
        try {
            val cardResponse = cardApiResponseAdapter.fromJson(malformedJson)
            assertNotNull("Card response should not be null", cardResponse)
            
            // Since score is typed as String?, invalid strings should be accepted
            assertEquals("Invalid string should be preserved", "invalid_number", cardResponse!!.payerMatch?.score)
            
            println("✅ SUCCESS: Invalid string preserved as-is: ${cardResponse.payerMatch?.score}")
            
        } catch (e: Exception) {
            println("❌ FAILED: Unexpected error: ${e.message}")
            throw e
        }
    }

    @Test
    fun testCardResponseFromSharedFixtures() {
        println("📁 Testing card response parsing from shared fixtures")
        
        // Test pending card (matching our shared fixture)
        val pendingCardJson = """
        {
            "card_id": "c53bcd29-26ec-4121-b090-cd381cae745f",
            "state": "pending",
            "created_at": "2025-06-18T02:26:24.578379+00:00",
            "deleted": false
        }
        """.trimIndent()
        
        try {
            val cardResponse = cardApiResponseAdapter.fromJson(pendingCardJson)
            assertNotNull("Pending card response should not be null", cardResponse)
            
            assertEquals("Card ID should match", "c53bcd29-26ec-4121-b090-cd381cae745f", cardResponse!!.cardId)
            assertEquals("State should be pending", CardState.pending, cardResponse.state)
            assertEquals("Deleted should be false", false, cardResponse.deleted)
            assertNotNull("Created at should not be null", cardResponse.createdAt)
            
            println("✅ Pending card parsed successfully")
            
        } catch (e: Exception) {
            println("❌ FAILED: Pending card parsing failed: ${e.message}")
            throw e
        }
    }

    @Test
    fun testCardResponseWithPayerMatch() {
        println("💳 Testing card response with payer matching")
        
        val cardWithPayerJson = """
        {
            "card_id": "c1b93738-ddc0-4beb-9936-1f93fe0e4279",
            "state": "completed",
            "created_at": "2025-06-18T02:26:24.578379+00:00",
            "deleted": false,
            "payer_match": {
                "cardscan_payer_id": "pay_8otorlr4",
                "cardscan_payer_name": "UNITEDHEALTHCARE",
                "score": "0.95"
            }
        }
        """.trimIndent()
        
        try {
            val cardResponse = cardApiResponseAdapter.fromJson(cardWithPayerJson)
            assertNotNull("Card with payer match should not be null", cardResponse)
            
            assertEquals("Card ID should match", "c1b93738-ddc0-4beb-9936-1f93fe0e4279", cardResponse!!.cardId)
            assertEquals("State should be completed", CardState.completed, cardResponse.state)
            
            assertNotNull("Payer match should not be null", cardResponse.payerMatch)
            assertEquals("Payer ID should match", "pay_8otorlr4", cardResponse.payerMatch?.cardscanPayerId)
            assertEquals("Payer name should match", "UNITEDHEALTHCARE", cardResponse.payerMatch?.cardscanPayerName)
            assertEquals("Score should match", "0.95", cardResponse.payerMatch?.score)
            
            println("✅ Card with payer match parsed successfully")
            
        } catch (e: Exception) {
            println("❌ FAILED: Payer match parsing failed: ${e.message}")
            throw e
        }
    }

    @Test
    fun testEligibilityResponseParsing() {
        println("🔍 Testing eligibility response parsing")
        
        val eligibilityJson = """
        {
            "eligibility_id": "93376802-779b-42ad-bfa3-d6e99d5a02c9",
            "state": "processing",
            "card_id": "529e865d-78c2-4f9e-aa9a-addedf642c88",
            "created_at": "2025-06-18T02:26:24.578379+00:00",
            "deleted": false
        }
        """.trimIndent()
        
        try {
            val eligibilityResponse = eligibilityApiResponseAdapter.fromJson(eligibilityJson)
            assertNotNull("Eligibility response should not be null", eligibilityResponse)
            
            assertEquals("Eligibility ID should match", "93376802-779b-42ad-bfa3-d6e99d5a02c9", eligibilityResponse!!.eligibilityId)
            assertEquals("State should be processing", EligibilityState.processing, eligibilityResponse.state)
            assertEquals("Card ID should match", "529e865d-78c2-4f9e-aa9a-addedf642c88", eligibilityResponse.cardId)
            assertEquals("Deleted should be false", false, eligibilityResponse.deleted)
            
            println("✅ Eligibility response parsed successfully")
            
        } catch (e: Exception) {
            println("❌ FAILED: Eligibility parsing failed: ${e.message}")
            throw e
        }
    }

    @Test
    fun testErrorCardResponse() {
        println("⚠️  Testing error card response")
        
        val errorCardJson = """
        {
            "card_id": "b7012e64-24c6-4f85-8410-adf36fe03e8a",
            "state": "error",
            "created_at": "2025-06-18T02:26:24.578379+00:00",
            "deleted": false,
            "error": {
                "type": "RejectedCard",
                "message": "Rejecting scan for invalid card",
                "code": "UKN"
            }
        }
        """.trimIndent()
        
        try {
            val cardResponse = cardApiResponseAdapter.fromJson(errorCardJson)
            assertNotNull("Error card response should not be null", cardResponse)
            
            assertEquals("Card ID should match", "b7012e64-24c6-4f85-8410-adf36fe03e8a", cardResponse!!.cardId)
            assertEquals("State should be error", CardState.error, cardResponse.state)
            assertEquals("Deleted should be false", false, cardResponse.deleted)
            
            assertNotNull("Error should not be null", cardResponse.error)
            assertEquals("Error type should match", "RejectedCard", cardResponse.error?.type)
            assertEquals("Error message should match", "Rejecting scan for invalid card", cardResponse.error?.message)
            
            println("✅ Error card response parsed successfully")
            
        } catch (e: Exception) {
            println("❌ FAILED: Error response parsing failed: ${e.message}")
            throw e
        }
    }

    @Test
    fun testSnakeCaseFieldNames() {
        println("🐍 Testing snake_case field names from API")
        
        val snakeCaseJson = """
        {
            "card_id": "test-123",
            "created_at": "2023-01-01T00:00:00Z",
            "payer_match": {
                "cardscan_payer_name": "TEST_PAYER",
                "score": "0.88"
            }
        }
        """.trimIndent()
        
        try {
            val cardResponse = cardApiResponseAdapter.fromJson(snakeCaseJson)
            assertNotNull("Snake case response should not be null", cardResponse)
            
            // Verify snake_case fields are properly mapped
            assertEquals("Card ID should be parsed from snake_case", "test-123", cardResponse!!.cardId)
            assertNotNull("Created at should be parsed", cardResponse.createdAt)
            assertEquals("Payer name should be parsed", "TEST_PAYER", cardResponse.payerMatch?.cardscanPayerName)
            assertEquals("Score should be parsed", "0.88", cardResponse.payerMatch?.score)
            
            println("✅ Snake case field parsing successful")
            
        } catch (e: Exception) {
            println("❌ FAILED: Snake case parsing failed: ${e.message}")
            throw e
        }
    }

    @Test
    fun testEnumSerialization() {
        println("🔢 Testing enum serialization")
        
        val states = listOf("pending", "processing", "completed", "error")
        
        states.forEach { stateString ->
            val jsonWithState = """
            {
                "card_id": "test-enum-$stateString",
                "state": "$stateString", 
                "created_at": "2023-01-01T00:00:00Z",
                "deleted": false
            }
            """.trimIndent()
            
            try {
                val cardResponse = cardApiResponseAdapter.fromJson(jsonWithState)
                assertNotNull("Card response for state $stateString should not be null", cardResponse)
                
                val expectedState = when(stateString) {
                    "pending" -> CardState.pending
                    "processing" -> CardState.processing  
                    "completed" -> CardState.completed
                    "error" -> CardState.error
                    else -> throw IllegalArgumentException("Unknown state: $stateString")
                }
                
                assertEquals("State should match for $stateString", expectedState, cardResponse!!.state)
                
            } catch (e: Exception) {
                println("❌ FAILED: Enum parsing failed for $stateString: ${e.message}")
                throw e
            }
        }
        
        println("✅ All enum states parsed correctly")
    }

    @Test
    fun testNullAndOptionalFieldHandling() {
        println("🔧 Testing null and optional field handling")
        
        val minimalJson = """
        {
            "card_id": "test-minimal",
            "state": "pending",
            "created_at": "2023-01-01T00:00:00Z",
            "deleted": false
        }
        """.trimIndent()
        
        try {
            val cardResponse = cardApiResponseAdapter.fromJson(minimalJson)
            assertNotNull("Minimal card response should not be null", cardResponse)
            
            assertEquals("Card ID should match", "test-minimal", cardResponse!!.cardId)
            assertEquals("State should be pending", CardState.pending, cardResponse.state)
            assertEquals("Deleted should be false", false, cardResponse.deleted)
            
            // Optional fields should be null
            assertNull("Payer match should be null", cardResponse.payerMatch)
            assertNull("Details should be null", cardResponse.details)
            assertNull("Error should be null", cardResponse.error)
            assertNull("Images should be null", cardResponse.images)
            
            println("✅ Optional field handling successful")
            
        } catch (e: Exception) {
            println("❌ FAILED: Optional field handling failed: ${e.message}")
            throw e
        }
    }
}